core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java

```diff
@@ -557,135 +566,176 @@ public void applyState(final Address sender, int topologyId, boolean pushTransfe
-      InboundTransferTask inboundTransfer = null;
+      InboundTransferTask inboundTransfer;
       synchronized (transferMapsLock) {
          List<InboundTransferTask> inboundTransfers = transfersBySegment.get(stateChunk.getSegmentId());
          if (inboundTransfers != null) {
             inboundTransfer = inboundTransfers.stream().filter(task -> task.getSource().equals(sender)).findFirst().orElse(null);
+         } else {
+            inboundTransfer = null;
          }
       }
       if (inboundTransfer != null) {
-         if (stateChunk.getCacheEntries() != null) {
-            doApplyState(sender, stateChunk.getSegmentId(), stateChunk.getCacheEntries());
-         }
-
-         inboundTransfer.onStateReceived(stateChunk.getSegmentId(), stateChunk.isLastChunk());
+         return doApplyState(sender, stateChunk.getSegmentId(), stateChunk.getCacheEntries())
+                   .thenAccept(
+                      v -> inboundTransfer.onStateReceived(stateChunk.getSegmentId(), stateChunk.isLastChunk()));
       } else {
          if (cache.wired().getStatus().allowInvocations()) {
             log.ignoringUnsolicitedState(sender, stateChunk.getSegmentId(), cacheName);
          }
       }
+      return CompletableFutures.completedNull();
    }
 
-   private void doApplyState(Address sender, int segmentId, Collection<InternalCacheEntry> cacheEntries) {
+   private CompletionStage<?> doApplyState(Address sender, int segmentId,
+                                           Collection<InternalCacheEntry> cacheEntries) {
+      if (cacheEntries == null || cacheEntries.isEmpty())
+         return CompletableFutures.completedNull();
+
       if (trace) log.tracef("Applying new state chunk for segment %d of cache %s from node %s: received %d cache entries",
-            segmentId, cacheName, sender, cacheEntries.size());
+                            segmentId, cacheName, sender, cacheEntries.size());
 
       // CACHE_MODE_LOCAL avoids handling by StateTransferInterceptor and any potential locks in StateTransferLock
       boolean transactional = transactionManager != null;
-      for (InternalCacheEntry e : cacheEntries) {
+      if (transactional) {
+         Object key = "N/A";
+         Transaction transaction = new ApplyStateTransaction();
+         InvocationContext ctx = icf.createInvocationContext(transaction, false);
+         LocalTransaction localTransaction = ((LocalTxInvocationContext) ctx).getCacheTransaction();
          try {
-            InvocationContext ctx;
-            if (transactional) {
-               transactionManager.begin();
-               ctx = icf.createInvocationContext(transactionManager.getTransaction(), true);
-               ((TxInvocationContext) ctx).getCacheTransaction().setStateTransferFlag(PUT_FOR_STATE_TRANSFER);
+            localTransaction.setStateTransferFlag(PUT_FOR_STATE_TRANSFER);
+            for (InternalCacheEntry e : cacheEntries) {
```

Comment:
I am not as picky about `putAll` here as we share the context and the entries aren't written to the store, however is there a benefit to it?

Reply:
It's the same limitation with `PutMapCommand` having a single `metadata` field for all the values.